plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,908 @@
+		Password:         o.Password,
+	}
+	provider, err := openstack.NewClient(authOption.IdentityEndpoint)
+	if err != nil {
+		return fmt.Errorf("unable to create client for OpenStack endpoint %v", err)
+	}
+
+	ctx := context.Background()
+	client, err := o.HTTPClientConfig.CreateClient(ctx)
+	if err != nil {
+		return err
+	}
+
+	provider.HTTPClient = *client
+
+	if err := openstack.Authenticate(provider, authOption); err != nil {
+		return fmt.Errorf("unable to authenticate OpenStack user %v", err)
+	}
+
+	// Create required clients and attach to the OpenStack struct
+	if o.identity, err = openstack.NewIdentityV3(provider, gophercloud.EndpointOpts{}); err != nil {
+		return fmt.Errorf("unable to create V3 identity client %v", err)
+	}
+
+	if err := o.gatherServices(); err != nil {
+		o.Log.Warnf("failed to get resource openstack services %v", err)
+	}
+
+	if o.compute, err = openstack.NewComputeV2(provider, gophercloud.EndpointOpts{}); err != nil {
+		return fmt.Errorf("unable to create V2 compute client %v", err)
+	}
+
+	// Create required clients and attach to the OpenStack struct
+	if o.network, err = openstack.NewNetworkV2(provider, gophercloud.EndpointOpts{}); err != nil {
+		return fmt.Errorf("unable to create V2 network client %v", err)
+	}
+
+	// The Orchestration service is optional
+	if o.ContainsService("orchestration") {
+		if o.stack, err = openstack.NewOrchestrationV1(provider, gophercloud.EndpointOpts{}); err != nil {
+			return fmt.Errorf("unable to create V1 stack client %v", err)
+		}
+	}
+
+	// The Cinder volume storage service is optional
+	if o.ContainsService("volumev2") {
+		if o.volume, err = openstack.NewBlockStorageV2(provider, gophercloud.EndpointOpts{}); err != nil {
+			return fmt.Errorf("unable to create V2 volume client %v", err)
+		}
+	}
+
+	return nil
+}
+
+// Gather gathers resources from the OpenStack API and accumulates metrics.  This
+// implements the Input interface.
+func (o *OpenStack) Gather(acc telegraf.Accumulator) error {
+	// Gather resources.  Note service harvesting must come first as the other
+	// gatherers are dependant on this information.
+	gatherers := map[string]func(telegraf.Accumulator) error{
+		"projects":      o.gatherProjects,
+		"hypervisors":   o.gatherHypervisors,
+		"flavors":       o.gatherFlavors,
+		"servers":       o.gatherServers,
+		"volumes":       o.gatherVolumes,
+		"storage_pools": o.gatherStoragePools,
+		"subnets":       o.gatherSubnets,
+		"ports":         o.gatherPorts,
+		"networks":      o.gatherNetworks,
+		"aggregates":    o.gatherAggregates,
+		"nova_services": o.gatherNovaServices,
+		"agents":        o.gatherAgents,
+		"stacks":        o.gatherStacks,
+	}
+
+	for _, service := range o.EnabledServices {
+		if service == "services" {
+			o.accumulateServices(acc)
+			continue
+		}
+		gatherer := gatherers[service]
+		if err := gatherer(acc); err != nil {
+			acc.AddError(fmt.Errorf("failed to get resource %q %v", service, err))
+		}
+	}
```

Comment:
The comment says
> Note service harvesting must come first as the other gatherers are dependant on this information.

However, the loop does not guarantee that `services` are accumulated first. I also assume that `services` always has to be "enabled" i.e. in `EnabledServices`, so you can do the following
```suggestion
	// Gather resources.  Note service harvesting must come first as the other
	// gatherers are dependant on this information.
	gatherers := map[string]func(telegraf.Accumulator) error{
		"projects":      o.gatherProjects,
		"hypervisors":   o.gatherHypervisors,
		"flavors":       o.gatherFlavors,
		"servers":       o.gatherServers,
		"volumes":       o.gatherVolumes,
		"storage_pools": o.gatherStoragePools,
		"subnets":       o.gatherSubnets,
		"ports":         o.gatherPorts,
		"networks":      o.gatherNetworks,
		"aggregates":    o.gatherAggregates,
		"nova_services": o.gatherNovaServices,
		"agents":        o.gatherAgents,
		"stacks":        o.gatherStacks,
	}

	o.accumulateServices(acc)
	for _, service := range o.EnabledServices {
		if service == "services" {
			continue
		}
		gatherer := gatherers[service]
		if err := gatherer(acc); err != nil {
			acc.AddError(fmt.Errorf("failed to get resource %q %v", service, err))
		}
	}
```
Or filter out `services` from the list of `EnabledServices` in `Init()`.

Reply:
There are two methods for 'services' one to gather resources(gatherServices function) and store in a map, other to output the data to accumulator(accumulateServices function). 'gatherServices' is called in Init() and depending on whether 'services' is in EnabledServices, accumulateServices method is called.

Reply:
Ok, can you please then add a comment to the `if service == "services" {` part explaining why you are doing it the way you do it?